### PR TITLE
Adding share backup with same name between users

### PR DESCRIPTION
**Pull request by sn-px**
What this PR does / why we need it:
Adding test for share backup between two users

Which issue(s) this PR fixes (optional)
Closes # PA-331

Logs : http://aetos.pwx.purestorage.com/resultSet/testSetID/87050

### CODE REVIEW

**Review by snigdha-px — APPROVED** [review]
LGTM

File: tests/backup/backup_test.go [hunk 1]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
 	})
 })
 
+// Share backup with created with same name between two users
+var _ = Describe("{SwapShareBackup}", func() {
+
+	// testrailID corresponds to: https://portworx.testrail.net/index.php?/cases/view/82940
+	numberOfUsers := 2
+	users := make([]string, 0)
+	userbackupLocationMapping := map[string]string{}
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6523]: Make it `userBackupLocationMapping`

File: tests/backup/backup_test.go [hunk 2]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+				for _, user := range users {
+					credName := fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+					backup.AddRoleToUser(user, backup.InfrastructureOwner, fmt.Sprintf("Adding Infra Owner role to %s", user))
+					NonAdminUserctx, err := backup.GetNonAdminCtx(user, "Password1")
+					log.FailOnError(err, fmt.Sprintf("Failed to fetch ctx for custom user: [%v]", err))
+					CreateCloudCredentialNonAdminUser(provider, credName, cloudCredUID, orgID, NonAdminUserctx)
+					credNames = append(credNames, credName)
+					time.Sleep(time.Minute * 1)
+					backupLocationName := fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+					CreateS3BackupLocationNonAdminUser(backupLocationName, backupLocationUID, credName, cloudCredUID, BucketName, orgID, "", NonAdminUserctx)
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6589]: Handle the error here

File: tests/backup/backup_test.go [hunk 3]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+			log.InfoD(fmt.Sprintf("Creating bucket, cloud credentials and backup location for %s and %s", users[0], users[1]))
+			for _, provider := range providers {
+				BucketName = fmt.Sprintf("%s-%v", provider, buckets[0])
+				cloudCredUID = uuid.New()
+				backupLocationUID = uuid.New()
+
+				for _, user := range users {
+					credName := fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+					backup.AddRoleToUser(user, backup.InfrastructureOwner, fmt.Sprintf("Adding Infra Owner role to %s", user))
+					NonAdminUserctx, err := backup.GetNonAdminCtx(user, "Password1")
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6583]: We have conventionally names this as `ctxNonAdmin`. See if you can change this

File: tests/backup/backup_test.go [hunk 4]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+				}
+
+			}
+		})
+		for _, user := range users {
+			Step(fmt.Sprintf("Register source and destination cluster for backup on %s ", user), func() {
+				log.InfoD("Registering Source and Destination clusters as user : %s and verifying the status", user)
+				ctx, err := backup.GetNonAdminCtx(user, "Password1")
+				log.FailOnError(err, "Fetching %s ctx", user)
+				CreateSourceAndDestClusters(orgID, "", "", ctx)
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6600]: Handle the error here

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+				time.Sleep(time.Minute * 2)
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6601]: Remove `time.Sleep` here. Not needed.

File: tests/backup/backup_test.go [hunk 6]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+				dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, "Verifying backup cluster status")
+			})
+			Step(fmt.Sprintf("Taking backup of applications as %s", user), func() {
+				log.InfoD("Taking backup of applications as user : %s ", user)
+				backupName = "backup1"
+				ctx, err := backup.GetNonAdminCtx(user, "Password1")
+				log.FailOnError(err, "Fetching testuser ctx")
+				err = CreateBackup(backupName, SourceClusterName, userbackupLocationMapping[user], backupLocationUID, []string{bkpNamespaces[0]},
+					nil, orgID, clusterUid, "", "", "", "", ctx)
+				dash.VerifyFatal(err, nil, "Verifying backup creation")
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6612]: Mentioned the backup name here

File: tests/backup/backup_test.go [hunk 7]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+				ctx, err := backup.GetNonAdminCtx(user, "Password1")
+				log.FailOnError(err, "Fetching %s ctx", user)
+				CreateSourceAndDestClusters(orgID, "", "", ctx)
+				time.Sleep(time.Minute * 2)
+				clusterStatus, clusterUid = Inst().Backup.RegisterBackupClusterNonAdminUser(orgID, SourceClusterName, "", ctx)
+				dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, "Verifying backup cluster status")
+			})
+			Step(fmt.Sprintf("Taking backup of applications as %s", user), func() {
+				log.InfoD("Taking backup of applications as user : %s ", user)
+				backupName = "backup1"
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6607]: I would suggest giving a backup name which is unique and easily recognisable. Like `backup-<testrailID>`

File: tests/backup/backup_test.go [hunk 8]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+			backupCount := 0
+			for _, backup := range userBackups{
+				if backup == backupName {
+					backupCount = backupCount + 1
+				}
+			}
+			dash.VerifyFatal(backupCount, 2, fmt.Sprintf("Validating the shared backup [%s] is present in user context [%s]", backupName, users[0]))
+		})
+	})
+	JustAfterEach(func() {
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6652]: There is no cleanup for backups and backup location

File: tests/backup/backup_test.go [hunk 9]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+		})
+		Step(fmt.Sprintf("Share backup with %s", users[1]), func() {
+			log.InfoD(fmt.Sprintf("Share backup from %s to %s and validate", users[1], users[0]))
+			ctx, err := backup.GetNonAdminCtx(users[1], "Password1")
+
+			// Share backup with the user
+			err = ShareBackup(backupName, nil, []string{users[0]}, RestoreAccess, ctx)
+			log.FailOnError(err, "Failed to share backup %s", backupName)
+		})
+		Step(fmt.Sprintf("validate the backup shared %s is present in user context %s", backupName, users[0]), func() {
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6641]: You have to validate that the shared backup is actually restorable

File: tests/backup/backup_test.go [hunk 10]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+			ctx, err := backup.GetNonAdminCtx(users[0], "Password1")
+
+			// Share backup with the user
+			err = ShareBackup(backupName, nil, []string{users[1]}, RestoreAccess, ctx)
+			log.FailOnError(err, "Failed to share backup %s", backupName)
+		})
+		Step(fmt.Sprintf("validate the backup shared %s is present in user context %s", backupName, users[1]), func() {
+			userBackups, _ := GetAllBackupsForUser(users[1], "Password1")
+			backupCount := 0
+			for _, backup := range userBackups{
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6626]: Formatting - space after `userBackups`

File: tests/backup/backup_test.go [hunk 11]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+			ctx, err := backup.GetNonAdminCtx(users[1], "Password1")
+
+			// Share backup with the user
+			err = ShareBackup(backupName, nil, []string{users[0]}, RestoreAccess, ctx)
+			log.FailOnError(err, "Failed to share backup %s", backupName)
+		})
+		Step(fmt.Sprintf("validate the backup shared %s is present in user context %s", backupName, users[0]), func() {
+			userBackups, _ := GetAllBackupsForUser(users[1], "Password1")
+			backupCount := 0
+			for _, backup := range userBackups{
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:6644]: Formatting - space after `userBackups`

File: tests/backup/backup_test.go (hunk 8 above)
Comment (sn-px) [reply tests/backup/backup_test.go:6652]: no clean up required is required here as the location is created in user context , and user is wiped out in the clean up

File: tests/backup/backup_test.go (hunk 1 above)
Comment (sn-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go (hunk 2 above)
Comment (sn-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go [hunk 12]
```diff
@@ -6514,6 +6514,175 @@ var _ = Describe("{IssueMultipleDeletesForSharedBackup}", func() {
...
+		Step(fmt.Sprintf("Adding Credentials and Registering Backup Location for %s and %s", users[0], users[1]), func() {
+			log.InfoD(fmt.Sprintf("Creating bucket, cloud credentials and backup location for %s and %s", users[0], users[1]))
+			for _, provider := range providers {
+				BucketName = fmt.Sprintf("%s-%v", provider, buckets[0])
+				cloudCredUID = uuid.New()
+				backupLocationUID = uuid.New()
+
+				for _, user := range users {
+					credName := fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+					backup.AddRoleToUser(user, backup.InfrastructureOwner, fmt.Sprintf("Adding Infra Owner role to %s", user))
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go (hunk 4 above)
Comment (sn-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go (hunk 5 above)
Comment (sn-px) [reply tests/backup/backup_test.go]: removed

File: tests/backup/backup_test.go (hunk 10 above)
Comment (sn-px) [reply tests/backup/backup_test.go]: done

File: tests/backup/backup_test.go (hunk 11 above)
Comment (sn-px) [reply tests/backup/backup_test.go]: done

**Pull request closed without merging**